src/vma/dev/cq_mgr.cpp

```diff
@@ -123,15 +124,30 @@ cq_mgr::cq_mgr(ring_simple* p_ring, ib_ctx_handler* p_ib_ctx_handler, int cq_siz
 		configure(cq_size);
 }
 
-void cq_mgr::configure(int cq_size)
+void cq_mgr::configure(uint32_t *cq_size)
 {
 	vma_ibv_cq_init_attr attr;
 	memset(&attr, 0, sizeof(attr));
 
 	prep_ibv_cq(attr);
 
 	m_p_ibv_cq = vma_ibv_create_cq(m_p_ib_ctx_handler->get_ibv_context(),
-			cq_size - 1, (void *)this, m_comp_event_channel, 0, &attr);
+			*cq_size - 1, (void *)this, m_comp_event_channel, 0, &attr);
+
+	if (!m_p_ibv_cq && safe_mce_sys().hypervisor == mce_sys_var::HYPER_MSHV && !strncmp(m_p_ib_ctx_handler->get_ibname(), "mlx4", 4)) {
+		// This is a workaround for an issue with cq creation of mlx4 devices on upstream-driver
+		// VMs over Windows Hypervisor.
+		static uint32_t max_upstream_cq_size = 8192;
+		if (*cq_size > max_upstream_cq_size) {
+			cq_logdbg("cq creation failed with cq_size of %d. retrying with size of %d",
+					*cq_size, max_upstream_cq_size);
+
+			*cq_size = max_upstream_cq_size;
```

Comment:
as far `cq_size` can be passed as environment variable please consider information message for a user.
See https://github.com/Mellanox/libvma/blob/master/src/vma/util/sys_vars.cpp#L829 or https://github.com/Mellanox/libvma/blob/master/src/vma/util/sys_vars.cpp#L806